tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java

```diff
@@ -0,0 +1,296 @@
+            String clientId = "admin-" + ADMIN_CLIENT_ID_SEQUENCE.getAndIncrement();
+            LogContext logContext = new LogContext("[LegacyAdminClient clientId=" + clientId + "] ");
+            Time time = Time.SYSTEM;
+            Metrics metrics = new Metrics(time);
+            Metadata metadata = new Metadata(
+                    CommonClientConfigs.DEFAULT_RETRY_BACKOFF_MS,
+                    CommonClientConfigs.DEFAULT_RETRY_BACKOFF_MAX_MS,
+                    60 * 60 * 1000L, logContext,
+                    new ClusterResourceListeners());
+            metadata.bootstrap(ClientUtils.parseAndValidateAddresses(
+                    config.getList(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG),
+                    config.getString(CommonClientConfigs.CLIENT_DNS_LOOKUP_CONFIG)));
+            Selector selector = new Selector(
+                    DEFAULT_CONNECTION_MAX_IDLE_MS,
+                    metrics,
+                    time,
+                    "admin",
+                    ClientUtils.createChannelBuilder(config, time, logContext),
+                    logContext);
+            NetworkClient networkClient = new NetworkClient(
+                    selector,
+                    metadata,
+                    clientId,
+                    DEFAULT_MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION,
+                    DEFAULT_RECONNECT_BACKOFF_MS,
+                    DEFAULT_RECONNECT_BACKOFF_MAX,
+                    DEFAULT_SEND_BUFFER_BYTES,
+                    DEFAULT_RECEIVE_BUFFER_BYTES,
+                    config.getInt(CommonClientConfigs.REQUEST_TIMEOUT_MS_CONFIG),
+                    config.getLong(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG),
+                    config.getLong(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG),
+                    time,
+                    true,
+                    new ApiVersions(),
+                    logContext,
+                    MetadataRecoveryStrategy.NONE);
+            ConsumerNetworkClient highLevelClient = new ConsumerNetworkClient(
+                    logContext,
+                    networkClient,
+                    metadata,
+                    time,
+                    config.getLong(CommonClientConfigs.RETRY_BACKOFF_MS_CONFIG),
+                    config.getInt(CommonClientConfigs.REQUEST_TIMEOUT_MS_CONFIG),
+                    Integer.MAX_VALUE);
+            return new AdminClient(time, highLevelClient, metadata.fetch().nodes());
+        }
+
+        AdminClient(Time time, ConsumerNetworkClient client, List<Node> bootstrapBrokers) {
+            this.time = time;
+            this.client = client;
+            this.bootstrapBrokers = bootstrapBrokers;
+        }
+
+        private AbstractResponse send(Node target, AbstractRequest.Builder<?> request) {
+            RequestFuture<ClientResponse> future = client.send(target, request);
+            while (!future.isDone()) {
```

Comment:
Should we handle the timeout manually?

Reply:
I add a test case `testAdminSendNoBlock` to confirm `client.send` doesn't block forever. Thank you.